### DISCUSSION TITLE
Sneak Peek at "Night Driver Desktop"

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#651] [Announcements]
I've been tinkering off and on with Night Driver Desktop, which is a C# desktop app used to control multiple instances of NightDriverStrip (or Mesmerizer or NightDriver Pi... anything that can receive and process ND wifi packets).

A "Location" is a group of strips.  Since you can only hang about 1000 LEDs off an ESP32 at a reasonable framerate on a single channel, my under-eave Cabana lights, for example, are broken up into four controllers.  So are my shop cupboards.  

The effect draws across the entire location, then breaks up the pieces that go to each different strip, and since they're time-synced, it acts like one big canvas.

It can also drive matrices, as shown in this screen shot.  If anyone is interesting in working on or contributing to the C# WinForms app, please let me know!  There's also a lot of multithreaded C# plumbing that does the communication with the ESP32s, but it's largely all working well.

Cheers,
Dave
<img width="1077" alt="2024-09-23_17-12-43" src="https://github.com/user-attachments/assets/f1001d60-a093-421c-b093-c03de80f8a63">

**Comment by robertlipe** [comment]
Neat. I was thinking about something less ambitious than that for some of
my own home lighting once I realized I was in the land of insanity teaching
NightDriver about ESP-Now. Let a Real Computer fill a pixbuf and then
delegate out timestamped buffers to the ESP32s sprinkled around.

https://github.com/BradyMeighan/WLED-Studio, https://www.ledfx.app (not
quite the same space)  and https://github.com/isra17/wled-tools/tree/main
have all been on my list  of things to look at and seems to play in a
similar space. (There's another, more established peer to it in a bookmark
list somewhere (where ideas go to die ...).

WLED is actually close enough to our pixel buffer format  (hardly a
coincidence - an array of colors is an array of colors, though I'm sure
that SOMEONE tried to patent it) but with some little nuisances like
differing alignment or swapped x and y or something that's trivial that I
actually thought about implementing wire compatibility with WLED just to
use the numerous tools like that. I remember thinking they were close
enough that either could be the pattern leader and the others actually
managing pixel strands be the pixel followers without major damage in
either code base.

**Comment by davepl** [comment]
WLED is not really comparable, or I would use it.  Unless they’ve added canvases spanning multiple IP address with millisecond level timing sync…

You could build a Jumbotron with NightDriver where each row of the TV is 8 wide.  You could have 12 rows, each connected to its own Pi, receiving color data.  And the sync should prevent tearing.  I just don’t have that many to test with :-)

Going the OTHER way, you could certainly add code (and depending on the license, could just use their code) to add the ability to process WLED data to NightDriver.  That could be quite handy.

If you’re familiar with the WLED studio and the other app, and are up to speed on C#, let me know if you want to start tinkering on the app.  The core functionality works, you can even load and save definitions, but there’s TONs of work left to do, particularly with respect to scheduling and specifying effects, etc.

- Dave

**Comment by robertlipe** [comment]
I'll admit my network is clear enough and my displays small enough that UDP
jitter/loss isn't really an issue for me, but those that need it use DDP
<http://www.3waylabs.com/ddp/#Timecode%20Usage>,or DMX/E1.31 which all
handle display subsetting as well. I've experimented with WLED's UDP RT
sync (which has no time sync) and NightDriver's and really can't tell the
difference in *my* case. (Your case is surely more challenging. My
standards are waaaay lower than yours and I'm pretty sure I've seen only a
tiny fraction of your blinkies on cam.)  I think ARLS has a timecode, but
I'm not sure. MOST of those could be bolted into the NightDriver base
without a lot of suffering - see also: arrays of  RGB -  but at some point
it becomes WLED and I haven't exactly seen anyone asking for such things. I
was mostly flirting with this topic a few weeks ago as described in a
moment.

>  up to speed on C#,

Nope.

>  particularly with respect to scheduling and specifying effects,

Relevant to my interests (that's why I have bookmarks of those other apps
-  I was studying either implementing it for NightDriver or just using
something else) but unless I can accidentally program in C# (Which could
happen... ISTR it looked like Java with a Pascal accent, so maybe I could
do remedial stuff in it...) I'll be very little help to you on that one,
Dave, but I dig watching projects grow!

Holiday lighting is entering my thoughts more this time of year.

RJL

**Comment by heidepiek** [comment]
![afbeelding](https://github.com/user-attachments/assets/4d63fe59-62b8-4076-a9b1-1d9f4b3563bc)
The nightdriver desktop looks nice. Using something like that in my mancave looks like a good idea to me. Unfortunately I can't help with programming, I'm not skilled enough for that. Find it quite something that I got it running (totally surprised with myself😳 ) Can't create a new file myself. Can only load your demo. Hmm well, my mancave doesn't have all those nice little projects that are in and around your garage. Hope over time that there will be a working version for us tinkerers.💪 

Greetz harrie✋ 



https://github.com/user-attachments/assets/f16a78bc-1910-48f2-bad9-61cf30da6eef
